PR: https://github.com/odoo/odoo/pull/212679

From: 2713876dbc70d3984e584a9037a2206dcda4e84a
From: Arnold Moyaux
Date: 2025-09-07 15:50:27

Breaking data model changes scores: del:3 + add:2, change matches:
-    procurement_group_id = fields.Many2one('procurement.group', 'Procurement Group', copy=False)
+    stock_reference_ids = fields.Many2many(
-    _inherit = 'procurement.group'
-    sale_id = fields.Many2one('sale.order', 'Sale Order')
+    sale_ids = fields.Many2many(

Total Changes: 160

[REF] stock, *: add reference

The model `procurement.group` is replaced by a more flexible one:
`stock.reference`. The first one was somehow not well understood and a
bit limited by its many2one links to other business model (Sales Order,
Purchase Order, ...). The reference is made of a name and many2many to
all the documents integrated with Inventory. It's purpose has been
shrank to the minimal "make links between object". For instance
opening the manufacturing order created for a MTO sale. The merge of
stock move is delegated to existing parameters (`sale_id`,
created_purchase_line_ids`, ...)

The `mrp.production` needed an extra effort to group them according to
inheritance relationships. This commit add a new object to group
productions together either in a backorder relation or in a child-parent
one. This allow for instance to have a Sale Order linked to the MTO
product of the sold product and the production them selves to be linked
to the SO *and* the potential child MOs.

The procurement group propagation, before set on the Stock Rule, allowing
the stock moves to be merged in an existing picking or not has been
removed and replaced (for the creation of purchase order only) by a new
parameter on `res.partner` allowing the group of Request for Quotation
base on the scheduled date of the potential merge candidate (never, same
week, same month, always). All other merge scenarios should use batches
and waves transfers.

closes odoo/odoo#212679

Task: 4279803
Related: odoo/enterprise#87170
Related: odoo/upgrade#8326
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale_stock/models/__init__.py
+++ b/addons/sale_stock/models/__init__.py
@@ -9,3 +9,4 @@ from . import res_users
 from . import sale_order
 from . import sale_order_line
 from . import stock
+from . import stock_reference

--- a/addons/sale_stock/models/sale_order.py
+++ b/addons/sale_stock/models/sale_order.py
@@ -44,7 +44,9 @@ class SaleOrder(models.Model):
         search='_search_late_availability',
         help="True if any related picking has late availability"
     )
-    procurement_group_id = fields.Many2one('procurement.group', 'Procurement Group', copy=False)
+    stock_reference_ids = fields.Many2many(
+        'stock.reference', 'stock_reference_sale_rel',
+        'sale_id', 'reference_id', string='References', copy=False)
     effective_date = fields.Datetime("Effective Date", compute='_compute_effective_date', store=True, help="Completion date of the first delivery order.")
     expected_date = fields.Datetime( help="Delivery date you can promise to the customer, computed from the minimum lead time of "
                                           "the order lines in case of Service products. In case of shipping, the shipping policy of "
@@ -172,9 +174,12 @@ class SaleOrder(models.Model):
             # TODO: Log a note on each down document
             deadline_datetime = values.get('commitment_date')
             for order in self:
-                order.order_line.move_ids.date_deadline = deadline_datetime or order.expected_date
+                moves = order.order_line.move_ids.filtered(
+                    lambda m: m.state not in ('done', 'cancel') and m.location_dest_id.usage == 'customer'
+                )
+                moves.date_deadline = deadline_datetime or order.expected_date
 
-        res = super(SaleOrder, self).write(values)
+        res = super().write(values)
         if values.get('order_line') and self.state == 'sale':
             for order in self:
                 to_log = {}
@@ -283,7 +288,7 @@ class SaleOrder(models.Model):
             picking_id = picking_id[0]
         else:
             picking_id = pickings[0]
-        action['context'] = dict(default_partner_id=self.partner_id.id, default_picking_type_id=picking_id.picking_type_id.id, default_origin=self.name, default_group_id=picking_id.group_id.id)
+        action['context'] = dict(default_partner_id=self.partner_id.id, default_picking_type_id=picking_id.picking_type_id.id, default_origin=self.name, default_reference_ids=self.stock_reference_ids[:-1].id)
         return action
 
     def _prepare_invoice(self):

--- a/addons/sale_stock/models/sale_order_line.py
+++ b/addons/sale_stock/models/sale_order_line.py
@@ -246,18 +246,19 @@ class SaleOrderLine(models.Model):
                 # Propagate deadline on related stock move
                 line.move_ids.date_deadline = line.order_id.date_order + timedelta(days=line.customer_lead or 0.0)
 
-    def _prepare_procurement_values(self, group_id=False):
+    def _prepare_procurement_values(self):
         """ Prepare specific key for moves or other components that will be created from a stock rule
         coming from a sale order line. This method could be override in order to add other custom key that could
         be used in move/po creation.
         """
-        values = super(SaleOrderLine, self)._prepare_procurement_values(group_id)
+        values = super()._prepare_procurement_values()
         self.ensure_one()
         # Use the delivery date if there is else use date_order and lead time
         date_deadline = self.order_id.commitment_date or self._expected_date()
         date_planned = date_deadline - timedelta(days=self.order_id.company_id.security_lead)
         values.update({
-            'group_id': group_id,
+            'origin': self.order_id.name,
+            'reference_ids': self.order_id.stock_reference_ids,
             'sale_line_id': self.id,
             'date_planned': date_planned,
             'date_deadline': date_deadline,
@@ -325,27 +326,22 @@ class SaleOrderLine(models.Model):
 
         return self.env['stock.move'].browse(outgoing_moves_ids), self.env['stock.move'].browse(incoming_moves_ids)
 
-    def _get_procurement_group(self):
-        return self.order_id.procurement_group_id
-
-    def _prepare_procurement_group_vals(self):
+    def _prepare_reference_vals(self):
         return {
             'name': self.order_id.name,
-            'move_type': self.order_id.picking_policy,
-            'sale_id': self.order_id.id,
-            'partner_id': self.order_id.partner_shipping_id.id,
+            'sale_ids': [(4, self.order_id.id)],
         }
 
     def _create_procurements(self, product_qty, procurement_uom, values):
         self.ensure_one()
-        return [self.env['procurement.group'].Procurement(
+        return [self.env['stock.rule'].Procurement(
             self.product_id, product_qty, procurement_uom, self._get_location_final(),
             self.product_id.display_name, self.order_id.name, self.order_id.company_id, values)]
 
     def _action_launch_stock_rule(self, *, previous_product_uom_qty=False):
         """
-        Launch procurement group run method with required/custom fields generated by a
-        sale order line. procurement group will launch '_run_pull', '_run_buy' or '_run_manufacture'
+        Launch procurement run method with required/custom fields generated by a
+        sale order line. procurement will launch '_run_pull', '_run_buy' or '_run_manufacture'
         depending on the sale order line product rule.
         """
         if self.env.context.get("skip_procurement"):
@@ -360,22 +356,11 @@ class SaleOrderLine(models.Model):
             if float_compare(qty, line.product_uom_qty, precision_digits=precision) == 0:
                 continue
 
-            group_id = line._get_procurement_group()
-            if not group_id:
-                group_id = self.env['procurement.group'].create(line._prepare_procurement_group_vals())
-                line.order_id.procurement_group_id = group_id
-            else:
-                # In case the procurement group is already created and the order was
-                # cancelled, we need to update certain values of the group.
-                updated_vals = {}
-                if group_id.partner_id != line.order_id.partner_shipping_id:
-                    updated_vals.update({'partner_id': line.order_id.partner_shipping_id.id})
-                if group_id.move_type != line.order_id.picking_policy:
-                    updated_vals.update({'move_type': line.order_id.picking_policy})
-                if updated_vals:
-                    group_id.write(updated_vals)
-
-            values = line._prepare_procurement_values(group_id=group_id)
+            references = line.order_id.stock_reference_ids
+            if not references:
+                self.env['stock.reference'].create(line._prepare_reference_vals())
+
+            values = line._prepare_procurement_values()
             product_qty = line.product_uom_qty - qty
 
             line_uom = line.product_uom_id
@@ -383,7 +368,7 @@ class SaleOrderLine(models.Model):
             product_qty, procurement_uom = line_uom._adjust_uom_quantities(product_qty, quant_uom)
             procurements += line._create_procurements(product_qty, procurement_uom, values)
         if procurements:
-            self.env['procurement.group'].run(procurements)
+            self.env['stock.rule'].run(procurements)
 
         # This next block is currently needed only because the scheduler trigger is done by picking confirmation rather than stock.move confirmation
         orders = self.mapped('order_id')

--- a/addons/sale_stock/models/stock.py
+++ b/addons/sale_stock/models/stock.py
@@ -3,7 +3,7 @@
 
 from collections import defaultdict
 
-from odoo import api, fields, models, _
+from odoo import api, fields, models, Command
 from odoo.tools.sql import column_exists, create_column
 
 
@@ -121,6 +121,13 @@ class StockMove(models.Model):
     def _get_all_related_sm(self, product):
         return super()._get_all_related_sm(product) | self.filtered(lambda m: m.sale_line_id.product_id == product)
 
+    def _prepare_procurement_values(self):
+        res = super()._prepare_procurement_values()
+        # to pass sale_line_id fom SO to MO in mto
+        if self.sale_line_id:
+            res['sale_line_id'] = self.sale_line_id.id
+        return res
+
 
 class StockMoveLine(models.Model):
     _inherit = "stock.move.line"
@@ -129,12 +136,6 @@ class StockMoveLine(models.Model):
         return 'customer' in {self.location_id.usage, self.location_dest_id.usage}
 
 
-class ProcurementGroup(models.Model):
-    _inherit = 'procurement.group'
-
-    sale_id = fields.Many2one('sale.order', 'Sale Order')
-
-
 class StockRule(models.Model):
     _inherit = 'stock.rule'
 
@@ -149,25 +150,33 @@ class StockPicking(models.Model):
 
     sale_id = fields.Many2one('sale.order', compute="_compute_sale_id", inverse="_set_sale_id", string="Sales Order", store=True, index='btree_not_null')
 
-    @api.depends('group_id')
+    @api.depends('reference_ids.sale_ids', 'move_ids.sale_line_id.order_id')
     def _compute_sale_id(self):
         for picking in self:
-            picking.sale_id = picking.group_id.sale_id
+            # picking and move should have a link to the SO to see the picking on the stat button.
+            # This will filter the move chain to the delivery moves only.
+            picking.sale_id = picking.move_ids.sale_line_id.order_id
+
+    @api.depends('move_ids.sale_line_id')
+    def _compute_move_type(self):
+        super()._compute_move_type()
+        for picking in self:
+            sale_orders = picking.move_ids.sale_line_id.order_id
+            if sale_orders:
+                if any(so.picking_policy == "direct" for so in sale_orders):
+                    picking.move_type = "direct"
+                else:
+                    picking.move_type = "one"
 
     def _set_sale_id(self):
-        if self.group_id:
-            self.group_id.sale_id = self.sale_id
+        if self.reference_ids and self.sale_id:
+            self.reference_ids.sale_ids = Command.link(self.sale_id)
         else:
             if self.sale_id:
-                vals = {
-                    'sale_id': self.sale_id.id,
+                self.env['stock.reference'].create({
+                    'sale_ids': [Command.link(self.sale_id.id)],
                     'name': self.sale_id.name,
-                }
-            else:
-                vals = {}
-
-            pg = self.env['procurement.group'].create(vals)
-            self.group_id = pg
+                })
 
     def _auto_init(self):
         """
@@ -181,6 +190,51 @@ class StockPicking(models.Model):
             create_column(self.env.cr, 'stock_picking', 'sale_id', 'int4')
         return super()._auto_init()
 
+    def _action_done(self):
+        res = super()._action_done()
+        sale_order_lines_vals = []
+        for move in self.move_ids:
+            ref_sale = move.picking_id.reference_ids.sale_ids
+            sale_order = ref_sale and ref_sale[0] or move.sale_line_id.order_id
+            # Creates new SO line only when pickings linked to a sale order and
+            # for moves with qty. done and not already linked to a SO line.
+            if not sale_order or move.sale_line_id or not move.picked or not (
+                (move.location_dest_id.usage in ['customer', 'transit'] and not move.move_dest_ids)
+                or (move.location_id.usage == 'customer' and move.to_refund)
+            ):
+                continue
+            product = move.product_id
+            quantity = move.quantity
+            if move.location_id.usage in ['customer', 'transit']:
+                quantity *= -1
+
+            so_line_vals = {
+                'move_ids': [(4, move.id, 0)],
+                'name': product.display_name,
+                'order_id': sale_order.id,
+                'product_id': product.id,
+                'product_uom_qty': 0,
+                'qty_delivered': quantity,
+                'product_uom_id': move.product_uom.id,
+            }
+            so_line = sale_order.order_line.filtered(lambda sol: sol.product_id == product)
+            if product.invoice_policy == 'delivery':
+                # Check if there is already a SO line for this product to get
+                # back its unit price (in case it was manually updated).
+                if so_line:
+                    so_line_vals['price_unit'] = so_line[0].price_unit
+            elif product.invoice_policy == 'order':
+                # No unit price if the product is invoiced on the ordered qty.
+                so_line_vals['price_unit'] = 0
+            # New lines should be added at the bottom of the SO (higher sequence number)
+            if not so_line:
+                so_line_vals['sequence'] = max(sale_order.order_line.mapped('sequence')) + len(sale_order_lines_vals) + 1
+            sale_order_lines_vals.append(so_line_vals)
+
+        if sale_order_lines_vals:
+            self.env['sale.order.line'].with_context(skip_procurement=True).create(sale_order_lines_vals)
+        return res
+
     def _log_less_quantities_than_expected(self, moves):
         """ Log an activity on sale order that are linked to moves. The
         note summarize the real processed quantity and promote a

--- a/None
+++ b/addons/sale_stock/models/stock_reference.py
@@ -0,0 +1,9 @@
+from odoo import fields, models
+
+
+class StockReference(models.Model):
+    _inherit = 'stock.reference'
+
+    sale_ids = fields.Many2many(
+        'sale.order', 'stock_reference_sale_rel', 'reference_id',
+        'sale_id', string="Sales")
